### ISSUE TITLE
cache utils

### ISSUE THREAD

**Issue by schleichardt** [sphereio/sphere-jvm-sdk#429]
This is related to https://github.com/sphereio/sphere-sunrise/issues/58 .

A shop should be able to cache common things and may rebuild them.

I think we should remove `void rebuildAsync()` from the CategoryTree interface, since the default implementation does nothing about this.

I suggest to create the following interfaces (a lot of changes in the old interface):

```java
//in sphere models artifact included
public interface CategoryTree {

List<Category> getRoots();
Optional<Category> findById(String id);
Optional<Category> findBySlug(Locale locale, String slug);
List<Category> getAllAsFlatList();
List<Category> getRoots();

}
```

in a utils module which depends on models and the Java client:
```java
public interface Cached<T> {

void rebuildAsync();
T get();

}
```

```java
public interface CachedCategoryTree extends Cached<CategoryTree> {
void rebuildAsync();
CategoryTree get();
}
```

```java
public class CachedCategoryTreeImpl extends CachedImpl<CategoryTree> {
@Inject
public CachedCategoryTreeImpl(final SphereClient client){/* ... */}
}
```

DI binding:
```java
bind(CachedCategoryTree.class).to(CachedCategoryTreeImpl.class).eager();//for prod
```

**Comment by schleichardt** [comment]
@lauraluiz What do you think?

**Comment by lauraluiz** [comment]
Yes, the caching part should not only be for Categories and it should not be directly in the JVM SDK Models, but rather depending on it.

But regarding the interface Cached, we probably have to think carefully what to put in there.
For example, do we need to manually trigger `rebuildAsync` every certain time, or maybe is the module itself that builds when the cache is outdated?
Basically, what kind of caching module do we want to build? A dummy or a smart one?

**Comment by schleichardt** [comment]
As shop developer you can use whatever you want to cache, but it would be nice to provide some basic stuff. So yes, for rebuildAsync no scheduled execution is planned. It should be already enough to have a cache aware container and to provide an endpoint to refresh it.

**Comment by schleichardt** [comment]
So if you don't see an urgent need for Sunrise, I would move this to the post 1.0.0 backlog.

**Comment by lauraluiz** [comment]
It won't be in the MVP of Sunrise, so you can move it post 1.0.0.

**Comment by schleichardt** [comment]
see https://github.com/sphereio/sphere-jvm-sdk/pull/930 concerning product types
